### DISCUSSION TITLE
Version 14.0, using directive for input fields does not work, is it wrong or is it not supported?

### DISCUSSION THREAD

**Discussion by tianlongJ** [graphql-java/graphql-java#3622] [General]
example：
directive @dateFormat(pattern:String) on INPUT_FIELD_DEFINITION

input Test {
date: Date @dateFormat(pattern: "yyyy")
}

public class DateFormatDirective implements SchemaDirectiveWiring {
    public DateFormatDirective() {
    }

    public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
        GraphQLFieldDefinition field = (GraphQLFieldDefinition)environment.getElement();
        DataFetcher originalDataFetcher = (DataFetcher)ClassUtil.call("getDataFetcher", field);
        DataFetcher dataFetcher = DataFetcherFactories.wrapDataFetcher(originalDataFetcher, (dataFetchingEnvironment, value) -> {
            String format = (String)dataFetchingEnvironment.getArgument("format");
            if (StringUtil.isBlank(format)) {
                if (value instanceof LocalDateTime) {
                    return Date.from(((LocalDateTime)value).atZone(ZoneId.systemDefault()).toInstant()).getTime();
                } else {
                    return value instanceof Date ? value : value;
                }
            } else {
                DateTimeFormatter dateTimeFormatter = this.buildFormatter(format);
                if (value instanceof LocalDateTime) {
                    return dateTimeFormatter.format((LocalDateTime)value);
                } else {
                    return value instanceof Date ? dateTimeFormatter.format(LocalDateTime.ofInstant(((Date)value).toInstant(), ZoneId.systemDefault())) : value;
                }
            }
        });
        return field.transform((builder) -> {
            builder.argument(GraphQLArgument.newArgument().name("format").type(Scalars.GraphQLString)).dataFetcher(dataFetcher);
        });
    }

    @Override
    public GraphQLInputObjectField onInputObjectField(SchemaDirectiveWiringEnvironment<GraphQLInputObjectField> environment) {
        GraphQLInputObjectField field = environment.getElement();
        Object value = field.getDefaultValue();
        if (value != null) {
            Object pattern = field.getDirective("dateFormat").getArgument("format").getValue();
            String input = value.toString();
            value = LocalDateTime.parse(input, this.buildFormatter(Optional.ofNullable(pattern).map(Object::toString).orElse(null)));
        }
        Object finalValue = value;
        return field.transform(builder -> builder.defaultValue(finalValue).build());
    }

    private DateTimeFormatter buildFormatter(String format) {
        String dtFormat = format != null ? format : "YYYY-MM-dd'T'HH:mm:ss.SSS'Z'";
        return DateTimeFormatter.ofPattern(dtFormat);
    }
}

I want to format a string as a date type according to the content in the pattern. I tried using DateFormatDirective, but it didn't work. Neither onField nor onInputObjectField worked.

**Comment by bbakerman** [comment]
The first thingI will say is that V14.0 of graphql-java came out in Jan 202 - thats 4.5 years ago.  I dont think you will get good outcomes using it in 2024 and I dont think we can really help give great answers that will work with it because it so old.

https://github.com/graphql-java/graphql-java/releases/tag/v14.0

I urge you to update.

Are you trying to say that you want to format incoming data differently on input or format values on output.

Its not clear to me.

**Reply by tianlongJ** [reply]
I think I have found the problem. It is the problem of graphql tools package. The lower version does not have the function of custom directive. The schema object is created by it.
